File: backend/app/schemas/document.py
from __future__ import annotations

from datetime import datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict, computed_field

from app.models.enums import DocumentStatus, OCRProviderEnum, PreprocessProfileEnum, ProcessingStage


class DocumentUploadResponse(BaseModel):
    document_id: str
    job_id: str
    status: DocumentStatus
    original_filename: str
    page_count: int
    # True when this upload's content-hash matched an already-processed,
    # non-deleted document and `force` was not set (spec section 9) --
    # `document_id`/`job_id` then point at that EXISTING document, not a
    # new one. Defaults False so existing callers that don't read this
    # field see unchanged behavior.
    is_duplicate: bool = False


class DocumentRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    original_filename: str
    file_extension: str
    mime_type: str
    file_size_bytes: int
    page_count: int
    status: DocumentStatus
    ocr_provider: OCRProviderEnum
    dpi: int
    preprocess_profile: PreprocessProfileEnum
    error_message: Optional[str] = None
    created_at: datetime
    updated_at: datetime


class DocumentListResponse(BaseModel):
    items: list[DocumentRead]
    total: int
    page: int
    page_size: int


class DocumentStats(BaseModel):
    total: int
    processing: int
    completed: int
    failed: int


class ProcessRequest(BaseModel):
    ocr_provider: Optional[OCRProviderEnum] = None
    dpi: Optional[int] = None
    preprocess_profile: Optional[PreprocessProfileEnum] = None


class ProcessingJobRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    document_id: str
    status: DocumentStatus
    stage: ProcessingStage
    progress_percent: int
    ocr_provider: OCRProviderEnum
    dpi: int
    preprocess_profile: PreprocessProfileEnum
    started_at: Optional[datetime] = None
    finished_at: Optional[datetime] = None
    error_message: Optional[str] = None

    @computed_field  # type: ignore[prop-decorator]
    @property
    def processing_duration_seconds(self) -> Optional[float]:
        """Authoritative total pipeline duration -- the frontend timer
        ticks live off `started_at` while a job is active (a purely visual
        approximation), but this backend-computed value is what a
        completed/failed/cancelled job's frozen final display must use
        (never a value the frontend derived itself from timestamps that
        could be stale). `None` while still running (there is no final
        duration yet) or if the job never actually reached PROCESSING
        (`started_at` never set) -- matches `started_at`/`finished_at`
        already being set together on every terminal transition (see
        `document_service.update_job_progress`), so no separate stored
        column or migration is needed."""
        if self.started_at is None or self.finished_at is None:
            return None
        return max(0.0, (self.finished_at - self.started_at).total_seconds())


class ProgressEvent(BaseModel):
    document_id: str
    job_id: str
    stage: ProcessingStage
    percent: int
    page: Optional[int] = None
    message: str = ""
    status: DocumentStatus


class PageRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    page_number: int
    width: int
    height: int
    dpi: int
    rotation: float
    original_image_url: str
    processed_image_url: Optional[str] = None
    ocr_status: str
    layout_status: str
    table_status: str
    ocr_confidence_avg: Optional[float] = None
    layout_confidence_avg: Optional[float] = None
    table_confidence_avg: Optional[float] = None
    visual_similarity_score: Optional[float] = None


class QualityPageReport(BaseModel):
    page_number: int
    ocr_confidence_avg: Optional[float]
    layout_confidence_avg: Optional[float]
    table_confidence_avg: Optional[float]
    visual_similarity_score: Optional[float]
    needs_review: bool


class QualityReport(BaseModel):
    document_id: str
    pages: list[QualityPageReport]
    overall_visual_similarity: Optional[float]

File: frontend/src/types/document.ts
export type DocumentStatus =
  | "UPLOADED"
  | "QUEUED"
  | "PROCESSING"
  | "OCR_PROCESSING"
  | "LAYOUT_PROCESSING"
  | "TABLE_PROCESSING"
  | "RECONSTRUCTING"
  | "EXPORTING"
  | "COMPLETED"
  | "FAILED"
  | "CANCELLED";

export type ProcessingStage =
  | "upload"
  | "render"
  | "preprocess"
  | "ocr"
  | "layout"
  | "table"
  | "document_json"
  | "reconstruct"
  | "export"
  | "quality"
  | "done";

export type OCRProvider = "paddleocr" | "tesseract" | "nvidia" | "ollama";
export type PreprocessProfile = "FAST" | "BALANCED" | "HIGH_QUALITY";

export type LayoutBlockType =
  | "title"
  | "heading"
  | "subheading"
  | "paragraph"
  | "table"
  | "image"
  | "chart"
  | "signature"
  | "stamp"
  | "handwritten"
  | "header"
  | "footer"
  | "page_number"
  | "footnote"
  | "horizontal_line"
  | "vertical_line";

export interface BBox {
  x1: number;
  y1: number;
  x2: number;
  y2: number;
}

export interface DocumentRecord {
  id: string;
  original_filename: string;
  file_extension: string;
  mime_type: string;
  file_size_bytes: number;
  page_count: number;
  status: DocumentStatus;
  ocr_provider: OCRProvider;
  dpi: number;
  preprocess_profile: PreprocessProfile;
  error_message: string | null;
  created_at: string;
  updated_at: string;
}

export interface DocumentListResponse {
  items: DocumentRecord[];
  total: number;
  page: number;
  page_size: number;
}

export interface DocumentStats {
  total: number;
  processing: number;
  completed: number;
  failed: number;
}

export interface ProcessingJob {
  id: string;
  document_id: string;
  status: DocumentStatus;
  stage: ProcessingStage;
  progress_percent: number;
  ocr_provider: OCRProvider;
  dpi: number;
  preprocess_profile: PreprocessProfile;
  started_at: string | null;
  finished_at: string | null;
  error_message: string | null;
  // Authoritative total pipeline duration, computed backend-side --
  // null while the job hasn't started or is still running (the frontend
  // Process Timer ticks live off `started_at` for that case), populated
  // once `finished_at` is set (COMPLETED/FAILED/CANCELLED all set this
  // together -- see `document_service.update_job_progress`).
  processing_duration_seconds: number | null;
}

export interface ProgressEvent {
  document_id: string;
  job_id: string;
  stage: ProcessingStage;
  percent: number;
  page: number | null;
  message: string;
  status: DocumentStatus;
}

export interface PageRecord {
  id: string;
  page_number: number;
  width: number;
  height: number;
  dpi: number;
  rotation: number;
  original_image_url: string;
  processed_image_url: string | null;
  ocr_status: string;
  layout_status: string;
  table_status: string;
  ocr_confidence_avg: number | null;
  layout_confidence_avg: number | null;
  table_confidence_avg: number | null;
  visual_similarity_score: number | null;
}

export interface OCRBlockRecord {
  id: string;
  page_id: string;
  block_id: string;
  line_id: string;
  text: string;
  original_text: string;
  confidence: number;
  bbox: BBox;
  polygon: number[][];
  language: string;
  is_reviewed: boolean;
}

export interface LayoutBlockRecord {
  id: string;
  page_id: string;
  block_type: LayoutBlockType;
  bbox: BBox;
  confidence: number;
  z_order: number;
  content: { text?: string; word_ids?: string[] };
  style: Record<string, unknown>;
  is_edited: boolean;
  edited_text: string | null;
}

export interface TableCellRecord {
  id: string;
  table_id: string;
  row: number;
  column: number;
  rowspan: number;
  colspan: number;
  text: string;
  bbox: BBox;
  align_h: "LEFT" | "CENTER" | "RIGHT";
  align_v: "TOP" | "MIDDLE" | "BOTTOM";
  confidence: number;
  is_header: boolean;
  is_edited: boolean;
}

export interface TableRecord {
  id: string;
  page_id: string;
  bbox: BBox;
  n_rows: number;
  n_cols: number;
  confidence: number;
  detection_method: string;
  column_widths: number[];
  row_heights: number[];
  border_style: Record<string, unknown>;
  cells: TableCellRecord[];
}

export interface QualityPageReport {
  page_number: number;
  ocr_confidence_avg: number | null;
  layout_confidence_avg: number | null;
  table_confidence_avg: number | null;
  visual_similarity_score: number | null;
  needs_review: boolean;
}

export interface QualityReport {
  document_id: string;
  pages: QualityPageReport[];
  overall_visual_similarity: number | null;
}

// Library accession register (cumulative Master Excel) -- see
// backend/app/services/accession_service.py.
export interface AccessionSummary {
  total_records: number;
  needs_review_count: number;
  latest_record_date: string | null;
  total_documents_processed: number;
}
